### ISSUE TITLE
No errors when using `@abstractmethod` on non-ABC class or when instantiating such class

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#2547]
**To Reproduce**
```python3
# pyright: strict

from abc import abstractmethod

class MyAbstract: # does not inherit from ABC
    @abstractmethod # no error
    def do_something(self):
        pass

c = MyAbstract().do_something() # also no error
```

**Expected behavior**
An error report on the usage of `@abstractmethod` or at the instantiation of `MyAbstract`. Mypy for example errors out with `error: Cannot instantiate abstract class "MyAbstract" with abstract attribute "do_something"`.

**VS Code extension or command-line**
Pyright extension version 1.1.185

**Comment by erictraut** [comment]
The current behavior is intended. It matches the runtime behavior.

Detecting non-overridden abstract methods is computationally expensive at analysis time, so I wouldn't want to do it for classes that are not marked abstract (i.e. those that don't derive from `ABC` or some other class constructed from `ABCMeta`). If you want pyright to provide this check, you should mark the class as abstract.

There is perhaps an argument to be made for a warning or error when the `@abstractmethod` decorator is used within a non-abstract class. I'm wondering if there's a legitimate use of this. If not, why doesn't the decorator raise an exception at runtime when it's applied to a method in a non-abstract class?

**Comment by JelleZijlstra** [comment]
> why doesn't the decorator raise an exception at runtime when it's applied to a method in a non-abstract class?

I believe this would be difficult to do because the decorator runs before the class exists.

For what it's worth, I could imagine people doing this if they want to mark the method as abstract in a standardized way but have some reason to avoid the ABCMeta base class (e.g., runtime overhead or a metaclass conflict).

**Comment by oscarbenjamin** [comment]
> I could imagine people doing this if they want to mark the method as abstract in a standardized way but have some reason to avoid the ABCMeta base class (e.g., runtime overhead or a metaclass conflict).

Yes, absolutely. The `ABCMeta` metaclass has a number of awkward effects that should not really be connected to wanting to verify that abstract methods have been implemented in base classes. My experiments using `ABC` resulted in unacceptable slowdowns in macro benchmarks which I think is because of the runtime overhead that it adds to `isinstance`:
```python
In [1]: from abc import ABC

In [2]: class Abstract(ABC): pass

In [3]: class Concrete: pass

In [4]: a = Abstract()

In [5]: c = Concrete()

In [6]: %timeit isinstance(c, Concrete)
33.3 ns ± 0.161 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [7]: %timeit isinstance(a, Abstract)
34 ns ± 1.07 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [8]: %timeit isinstance(a, Concrete)
62.5 ns ± 0.711 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [10]: %timeit isinstance(c, Abstract)
478 ns ± 2.23 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```
Given the following mypy will report an error but pyright does not:
```python
from abc import abstractmethod

class A:
    @abstractmethod
    def foo(self) -> bool:
        ...

class B(A):
    ...

b = B()
```
Then:
```console
$ mypy t.py
t.py:11: error: Cannot instantiate abstract class "B" with abstract attribute "foo"  [abstract]
Found 1 error in 1 file (checked 1 source file)
$ pyright t.py 
0 errors, 0 warnings, 0 informations 
```

> Detecting non-overridden abstract methods is computationally expensive at analysis time, so I wouldn't want to do it for classes that are not marked abstract

The purpose of `abstractmethod` is to try to enforce that a  base class implements the method so I think that pyright should report this as an error by default. It is much better for this to be detected by a type checker than by the runtime mechanisms used by `ABCMeta`.

**Comment by oscarbenjamin** [comment]
A workaround here is:
```python
if TYPE_CHECKING:
    from abc import ABC
else:
    ABC = object
```
I do think though that pyright should report this as an error though. Literally the entire purpose of `@abstractmethod` is to prevent accidentally not implementing the method. The `abc` module predates type checking and so implements this in a bad way. If the module didn't already exist then it would be implemented today using decorators from the typing module rather than the weird runtime stuff that it currently does.

**Comment by oscarbenjamin** [comment]
Actually after playing around with this I see that `class Foo[E](object): pass` fails at runtime. This is a better workaround because it works with generic types:
```python
if TYPE_CHECKING:
    from abc import ABCMeta
else:
    ABCMeta = type

class Foo[E](metaclass=ABCMeta):
   ...
```
